custom_components/wnsm/sensor.py

```
@@ -157,6 +175,76 @@ def parse_quarterly_consumption_response(self, response):
 
+    async def _import_statistics(self, smartmeter: Smartmeter, start: datetime, sum_: Decimal):
+        """Import hourly consumption data into the statistics module, using start date and sum"""
+        if start.tzinfo is None:
+            raise ValueError("start datetime must be timezone-aware!")
+        # Have to be sure that full minutes are used. otherwise, the API returns a different
+        # interval
+        start = start.replace(minute=0, second=0, microsecond=0)
+
+        has_none = False
+        statistics = []
+        metadata = StatisticMetaData(
+            source="recorder",
+            statistic_id=f'sensor.{self.unique_id.lower()}',
+            name=self.name,
+            unit_of_measurement=self._attr_unit_of_measurement,
+            has_mean=False,
+            has_sum=True,
+        )
+        _LOGGER.debug(metadata)
+
+        now = datetime.now(timezone.utc).replace(minute=0, second=0, microsecond=0)
+        _LOGGER.debug(f"Selecting data up to {now}")
+        # FIXME: this loop is prone to endless loops, if the API returns something funny...
+        # Thus, we add a counter here as well. But there are probably better methods to prevent that
+        iterations = 50
+        while iterations > 0 and not has_none and start < now:
+            _LOGGER.debug(f"Select 24h of Data, using sum={sum_:.3f}, start={start}")
+            iterations -= 1
+            verbrauch = await self.get_consumption(smartmeter, start)
+            _LOGGER.debug(verbrauch)
+
+            # Check if this batch of data is valid and contains hourly statistics:
+            if not verbrauch.get('quarter-hour-opt-in'):
+                _LOGGER.warning(f"Data starting at {start} does not contain granular data! Opt-in was not set back then.")
+                start += timedelta(hours=24)  # Select the next day...
+                continue
+
+            if 'values' not in verbrauch:
+                _LOGGER.error("No values in API response!")
```

if it's an error, shouldn't we abort? or is it okay to lose some "days" worth of data?

---

True, Idk though how the API behaves. On an error, will it just not send any values (thus in this if you should raise an exception)?

---

I am also not quite sure how often this happens. What's your opinion on that:

on an error that is user specific (like wrong configuration, changed credentials, etc.) we should raise an exception.
on something that could be a flake or a hiccup on API or network side, we just log it as an error and proceed.

so in this case: maybe log an error and abort the update?

---

Yes, I think if there are no values, it is actually an API error.
Maybe it is also a good idea to add the [`raise_for_status`](https://requests.readthedocs.io/en/latest/api/#requests.Response.raise_for_status) after this line: https://github.com/reox/WienerNetzeSmartmeter/blob/6b6fab8919b953234a1ba9f3de89b25f4f05f9e8/custom_components/wnsm/api/client.py#L123
Because I think in such a case, the API will also return a 500